pkg/service/did/ion.go

```diff
@@ -88,6 +91,287 @@ func (i ionStoredDID) IsSoftDeleted() bool {
+			storedDID.DID.Services = newServices
+
+			currentPublicKeys, _ := collectPublicKeys(storedDID.DID)
+			newPublicKeys := newPubKeys(currentPublicKeys, request)
+			didDoc, err := addPublicKeysPatch(storedDID.DID, newPublicKeys)
+			if err != nil {
+				return nil, err
+			}
+			storedDID.DID = *didDoc
+
+			doc := ion.Document{
+				PublicKeys: newPublicKeys,
+				Services:   newServices,
+			}
+			recoveryPubKeyJWK, err := h.recoveryPublicKey(ctx, request.DID.String())
+			if err != nil {
+				return nil, err
+			}
+			storedDID.LongFormDID, err = ion.CreateLongFormDID(*recoveryPubKeyJWK, *state.preAnchor.nextUpdatePublicJWK, doc)
+			if err != nil {
+				return nil, errors.Wrap(err, "creating long form DID")
+			}
+			if err := StoreDID(ctx, storedDID, tx); err != nil {
+				return nil, errors.Wrap(err, "storing DID in storage")
+			}
+
+			state.preAnchor.storedDID = storedDID
+			if err := h.storeUpdateState(ctx, request.DID.String(), updateStates, tx); err != nil {
+				return nil, err
+			}
+
+			return state, nil
+		}, watchKeys)
+		if err != nil {
+			return nil, errors.Wrapf(err, "executing transition to %s", PreAnchorStatus)
+		}
+		state = execResp.(*updateState)
+	}
+
+	if state.status == PreAnchorStatus {
+		state.anchor = new(anchor)
+		if err := h.resolver.Anchor(ctx, state.preAnchor.updateOperation); err != nil {
+			// Signature errors are OK, as they mean that the update operation has already been applied. It means we haven't updated our updateKey to the latest one.
+			state.anchor.err = err
+			if !isSignatureError(err) {
+				state.status = AnchorErrorStatus
+				return nil, errors.Wrap(err, "anchoring update operation")
+			}
+		}
+		state.status = AnchoredStatus
+		if err := h.storeUpdateState(ctx, request.DID.String(), updateStates, h.storage.db); err != nil {
+			return nil, err
+		}
+	}
+
+	// TODO: wrap this all up in the watch key. Needs https://github.com/TBD54566975/ssi-service/pull/539 to be merged.
+	updateStoreRequest, err := keyToStoreRequest(updateKeyID(request.DID.String()), *state.preAnchor.nextUpdatePrivateJWK, request.DID.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting update private key to store request")
+	}
+	if err := h.keyStore.StoreKey(ctx, *updateStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion update private key")
+	}
+
+	if state.status == AnchoredStatus {
+		_, err = h.storage.db.Execute(ctx, func(ctx context.Context, tx storage.Tx) (any, error) {
+			state.status = DoneStatus
+			if err := h.storeUpdateState(ctx, request.DID.String(), updateStates, tx); err != nil {
+				return nil, err
+			}
+			return nil, nil
+		}, watchKeys)
+		if err != nil {
+			return nil, errors.Wrapf(err, "executing transition to %s", DoneStatus)
+		}
+	}
+
+	return &UpdateDIDResponse{
+		DID: state.preAnchor.storedDID.DID,
+	}, nil
+}
+
+func isSignatureError(_ error) bool {
+	// TODO: figure out how to determine this error from the body of the response.
```

Comment:
@decentralgabe could you help pointing me to the type of error that I would get if I send an operation to an ION node which has already been applied?